internal-dns-client/src/multiclient.rs

```diff
@@ -0,0 +1,651 @@
+                crate::Client::new(&format!("http://{}", addr), log.clone())
+            })
+            .collect::<Vec<_>>();
+
+        Self { log, clients }
+    }
+
+    /// Inserts all service records into the DNS server.
+    ///
+    /// This method is most efficient when records are sorted by SRV key.
+    pub async fn insert_dns_records(
+        &self,
+        records: &Vec<impl Service>,
+    ) -> Result<(), DnsError> {
+        let mut records = records.iter().peekable();
+
+        while let Some(record) = records.next() {
+            let srv = record.srv();
+            info!(self.log, "Inserting DNS record: {:?}", srv);
+
+            match &srv {
+                &crate::names::SRV::Service(_) => {
+                    let mut aaaa = vec![(record.aaaa(), record.address())];
+                    while let Some(record) = records.peek() {
+                        if record.srv() == srv {
+                            let record = records.next().unwrap();
+                            aaaa.push((record.aaaa(), record.address()));
+                        } else {
+                            break;
+                        }
+                    }
+
+                    self.insert_dns_records_internal(aaaa, srv).await?;
+                }
+                &crate::names::SRV::Backend(_, _) => {
+                    let aaaa = vec![(record.aaaa(), record.address())];
+                    self.insert_dns_records_internal(aaaa, record.srv())
+                        .await?;
+                }
+            };
+        }
+        Ok(())
+    }
+
+    // Utility function to insert:
+    // - A set of uniquely-named AAAA records, each corresponding to an address
+    // - An SRV record, pointing to each of the AAAA records.
+    async fn insert_dns_records_internal(
+        &self,
+        aaaa: Vec<(crate::names::AAAA, SocketAddrV6)>,
+        srv_key: crate::names::SRV,
+    ) -> Result<(), DnsError> {
+        let mut records = Vec::with_capacity(aaaa.len() + 1);
+
+        // Add one DnsKv per AAAA, each with a single record.
+        records.extend(aaaa.iter().map(|(name, addr)| DnsKv {
+            key: DnsRecordKey { name: name.to_string() },
+            records: vec![DnsRecord::Aaaa(*addr.ip())],
+        }));
+
+        // Add the DnsKv for the SRV, with a record for each AAAA.
+        records.push(DnsKv {
+            key: DnsRecordKey { name: srv_key.to_string() },
+            records: aaaa
+                .iter()
+                .map(|(name, addr)| {
+                    DnsRecord::Srv(Srv {
+                        prio: 0,
+                        weight: 0,
+                        port: addr.port(),
+                        target: name.to_string(),
+                    })
+                })
+                .collect::<Vec<_>>(),
+        });
+        self.dns_records_set(&records).await
+    }
+
+    /// Sets a records on all DNS servers.
+    ///
+    /// Returns an error if setting the record fails on any server.
```

Comment:
I think it's pretty likely we'll want to change this behavior to cope with one or more of the DNS servers failing.

However, I really want our behavior to be "fail loud" until we have a good way of coping. If one of the DNS servers starts failing, ideally Nexus could:
- Notice when it is actually setting the records, pick out which server failed
- If it continues to not respond (or throw errors), mark the service as "deprecated", re-establish service redundancy elsewhere

Until then though, if an internal DNS server fails, I want high-visibility on the error. Hopefully this tradeoff makes sense?

Reply:
Also of note: This is exclusively on the "setting DNS records" side of things. The "getting DNS records" side should be able to cope with a server going offline already.

Reply:
I generally agree about failing loud, but it does seem slightly at odds with the availability-first ethos of DNS. This is probably good for now though, without more information about how the client-side will work.